docs: Leader Receiver Creator ADR (#1100)
kyma-project/telemetry-manager: docs/contributor/arch/012-leader-receiver-creator.md

# 12. Leader Receiver Creator

Date: 2024-05-21

## Status

Proposed

## Context

> As part of metrics collection, the Telemetry module must expose Kubernetes API server metrics. The corresponding OpenTelemetry (OTel) Collector receiver for this task is k8s_cluster receiver. However, there are some open questions:

Author: I just found a reference to it in ADR-11 and we call it Kubernetes Cluster Receiver. Also added the link

> * How can we ensure that the receiver operates in high availability mode? Should it be integrated into the metric agent (currently used only for collecting node-affine metrics), incorporated into the gateway (which typically runs multiple replicas), or deployed as a separate component?

Reviewer: ```suggestion
* How can we ensure that the receiver operates in high availability mode? Should it be integrated into the metric agent (which currently only collects node-affine metrics), incorporated into the gateway (which typically runs multiple replicas), or deployed as a separate component?
```

* How can we prevent sending duplicate metrics?

## Decision

> The easiest way to ensure high availability and prevent duplicate metrics is to use the leader election pattern. In this way, only one replica of OTel Collector will have an active instance of k8s_cluster_receiver running. The rest

Reviewer: ```suggestion
The easiest way to ensure high availability and prevent duplicate metrics is to use the leader election pattern. In this way, only one replica of OTel Collector will have an active instance of k8s_cluster_receiver running. 
```

> of the replicas will be in standby mode. If the active instance fails, the standby instance will take over.

Reviewer: ```suggestion
The rest of the replicas are in standby mode. If the active instance fails, the standby instance takes over.
```

The leader election pattern is well known in Kubernetes, the following building blocks are available:
* [Lease](https://kubernetes.io/docs/concepts/architecture/leases/)
* [Resource lock client](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection/resourcelock)
* [Leader election client](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection)

> One way to integrate leader election with k8s_cluster receiver is to bundle the receiver with the leader election logic. However, this approach tightly couples it with the concrete receiver. If we need the leader election logic in another context, we will have to reimplement it. Additionally, it will be challenging to contribute this combined implementation to the community since the receiver already has its own maintainers.

Author: Also fixed Receiver Creator and Leader Receiver Creator

An alternative approach is to create a separate component responsible for leader election that manages another arbitrary sub-receiver. This pattern is already used by the [receiver_creator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/receivercreator/README.md), which can instantiate other receivers at runtime based on observed endpoints matching a configured rule.

In this way, the new receiver, let's call it leader_receiver_creator, can be configured to create k8s_cluster receiver only if it is the leader. Here's the proposed API:

> yaml

Author: Will remove

> In this way, the new receiver (let's call it `Leader Receiver Creator`) can be configured to create Kubernetes Cluster Receiver only if the underlying instance is a leader. Here's the proposed API:

Reviewer: ```suggestion
In this way, the new receiver (let's call it "Leader Receiver Creator") can be configured to create Kubernetes Cluster Receiver only if the underlying instance is a leader. Here's the proposed API:
```
I think we no longer need to format this as code font

leader_receiver_creator:
  auth_type: serviceAccount

>   leader_election:

Reviewer 2: I think we would need some property for a lease / lock name so that the user can configure which receivers are competing to become the leader of a specific group.

Author: You are totally right, I forgot about it

    enabled: true
    lease_duration: 15s
    renew_deadline: 10s
    retry_period: 2s
  receiver:
    k8s_cluster:
      auth_type: serviceAccount
      node_conditions_to_report: [Ready, MemoryPressure]
      allocatable_types_to_report: [cpu, memory]
      metrics:
      k8s.container.cpu_limit:
      enabled: false
      resource_attributes:
      container.id:
      enabled: false
```

> For a draft implementationy see: github.com/skhalash/leaderreceivercreator.

Reviewer: ```suggestion
For a draft implementation, see [github.com/skhalash/leaderreceivercreator](https://github.com/skhalash/leaderreceivercreator).
```

> The new receiver can be deployed as part of either the metric agent or the metric gateway. However, it is preferable to run it within the metric gateway to avoid an additional network hop, as these metrics are not strictly node-affine.

Reviewer: Is the new receiver the Leader Receiver Creator, or the Kubernetes Cluster Receiver?
Maybe instead of "The new receiver", use the name of the specific component for clarity.

Author: Leader Receiver Creator

Author: Changed the wording

## Consequences

> The new leader_receiver_creator can be maintained as an internal project and later contributed to the community.

Reviewer: ```suggestion
The new Leader Receiver Creator can be maintained as an internal project and later contributed to the community.
```

It can be reused in other contexts where leader election is required. It can be bundled into our custom OTel Collector image.

Open questions:

> * How to handle multiple different receivers that must follow the leader election pattern?

Reviewer 2: Should be solved with my previous comment
